ファイルアップロード用のアクションを追加

a5ob7r opened the pull request:
# 概要

オンラインストレージにファイルをアップロードするためのアクションを追加する。

コントローラ内の認証認可の処理に関しては、他コントローラでも利用され得るので、concernとして切り出した。分量的にこの部分が本PRの議論の中心である。

a5ob7r committed:
Add file upload action

a5ob7r committed:
Add the authorization concern to authorize access tokens

a5ob7r committed:
Change a shared_example description into Japanese

a5ob7r committed:
Add a description to required parameter to a shared_examples

a5ob7r committed:
Refactor token string and option validations

a5ob7r committed:
Add a comment about the valid Authorization header format

a5ob7r committed:
Add some missing test cases about invalid access tokens

tk0miya commented on app/controllers/concerns/authorization.rb:
```diff
@@ -0,0 +1,62 @@
+
+  private
+
+  def token_string
+    @token_string ||= nil
+  end
+
+  def token_options
+    @token_options ||= nil
+  end
```
単に `attr_reader` でよさそうに見えました。

a5ob7r replied:
attr_reader の存在は知っていたのですが、ドキュメントを見る限り、これだとインスタンス変数が宣言されていない時はエラーが発生するのでよろしくないなと考えました。

https://docs.ruby-lang.org/ja/latest/method/Module/i/attr_reader.html

ですが、Rubyだと未定義のインスタンス変数を参照した場合はnilが返るんですね。知りませんでした。確かにこれだと attr_reader で宣言するだけでよさそうですね。

https://docs.ruby-lang.org/ja/latest/doc/spec=2fvariables.html
https://jp.quora.com/Rubyで未定義のローカル変数を呼び出すとNameErrorになり-未定

a5ob7r replied:
トークン文字列およびオプションにコントローラからアクセス可能な仕様にする意図はないので、これらのreaderは削除しました。

tk0miya commented on app/controllers/concerns/authorization.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def set_access_token
+    @access_token = AccessToken.find_by token: token_string
+  end
```
ここも同じく attr_reader でよさそう。

もしくはこういう風に実装して、初回アクセス時に find_by するというのもよくやります。

```suggestion
  def access_token
    @access_token ||= AccessToken.find_by token: token_string
  end
```

a5ob7r replied:
提案していただいた方法を利用するように変更しました。

tk0miya commented on app/controllers/concerns/authorization.rb:
```diff
@@ -0,0 +1,62 @@
+  include ActionController::HttpAuthentication::Token
+
+  included do
+    before_action :set_token_and_options, :validate_token_and_options
+    before_action :set_access_token, :authenticate_access_token
```
雑談?
before_action で複数のハンドラを並べるというのはあまり見かけないですね

a5ob7r replied:
こちらで見かけて採用しました。

```ruby
class ApplicationController < ActionController::Base
  before_action :set_account, :authenticate
```

https://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Token.html

今回は、一つの before_action で一つのエラーに対応するように並べました。1行目ならトークンの未送信でエラー、2行目なら不正なトークンでエラー、3行目と4行目ではトークンの権限不足によりエラー、といった具合です。

tk0miya commented on app/controllers/concerns/authorization.rb:
```diff
@@ -0,0 +1,62 @@
+    render problem: { error: 'insufficient_scope', scope: }, status: :forbidden
+  end
+
+  def authorize_access_token_as_readable
```
authorize という単語は目的語を取って、〜〜に許可を与えるという表現をします。
https://eow.alc.co.jp/search?q=authorize

この場合だと「アクセストークンに許可を与える」という命名になるような気がします。


before_action でチェックをしたいのは「読み取り権限があること」「書き込み権限があること」なので
`#authorized_as_readable`, `#authorized_as_writable` あたりにするのはどうでしょうか。
アクセストークンは手段であって、目的ではないのでメソッド名に入れないという発想です。

a5ob7r replied:
> #authorized_as_readable, #authorized_as_writable あたりにするのはどうでしょうか。

なんで過去形なんだろう、と思ったけど、認可はすでに与えられているかどうか、つまりは過去の出来事に対する確認だからなんですかね。それか現在完了かな。

対して認証は、誰かを特定する、つまり現在に対する操作であるので時制が異なるという理解でいいのかな。

tk0miya replied:
the current user is authorized as readable とか This controller requires the current user authorized as readable を表現していますね。過去分詞なのかな。
ちなみに、ここでは「認可されているかどうか」を確認しているので、自制は現在ですね。
(過去に認可されても、現在認可されているかどうかはわからないので、過去は重要ではないです)

現在形の authorize にしてしまうと認可を与える、つまりトークンを発行するという名前だと誤解されそうです。

※ 状態の確認っぽさを出したほうが、より動きがわかりやすいこともあって `require_` とか `ensure_` って prefix をつけるとさらによさそうですね。たとえば `#require_authorized_as_readable` とか。

a5ob7r replied:
> ※ 状態の確認っぽさを出したほうが、より動きがわかりやすいこともあって require_ とか ensure_ って prefix をつけるとさらによさそうですね。たとえば #require_authorized_as_readable とか。

こちらの名前に変更しました。

tk0miya commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,22 @@
+  include Authorization
+
+  def create
+    user = access_token.user
```
これまで作ったいくつかのアプリだと `#current_user` というメソッドを用意していましたね。
上でもコメントをしましたが、アプリケーションのレイヤーではアクセストークンは手段でしかなく、認証されたか/されていないか、権限はあるか/ないかなどが関心事なので、アクセストークンは隠蔽してしまったほうがメインのコードがわかりやすくなると思います。

アクセストークンそのものに関心があるのは、ログインログアウトまわりの controller と Authorization concern だけで十分だと思います。

a5ob7r replied:
ふむふむ、なるほど。以前に#current_userというメソッドを用意したときは、不要ではと指摘されましたが、その時とは関心事が違うということですね。修正します。

tk0miya replied:
うう…ごめん、以前の指摘を覚えていないので、矛盾したことを言っていたらごめん。

夕会でも触れたように、アクセストークンそのものに触れたいというのはアプリのうちごく一部のはずで、多くの controller では「認証されたユーザーはどれか」「どういう権限が付与されているのか」が関心ごとだと思います。
ですから、敢えてアクセストークンに触らせる必要はないのでは、というのが今の感覚です。

a5ob7r replied:
すみません、僕の誤解でした。#current_userへの指摘ではあったのですが、指摘は、その内容や呼び出す箇所についてでした。そして、最終的に僕が#current_userを削除したので、それを短絡的に削除への指摘として記憶してしまったようです。

https://github.com/sls-training/2023-API-training-server/pull/38#discussion_r1243543165

a5ob7r replied:
失礼しました。

tk0miya commented on app/models/access_token.rb:
```diff
@@ -29,4 +29,8 @@ def revoked?
     update revoked_at: Time.current if revoked_at.nil?
   end
+
+  def authorize?(required_scope)
```
authorize は動詞なので、状態を確認するなら authorize *d* ? のほうがよさそう。
```suggestion
  def authorized?(required_scope)
```

a5ob7r replied:
それもそうですね、修正します。

tk0miya commented on config/routes.rb:
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :items, as: :files, only: %i[create]
```
controller の名前は ItemsController がいいんですかねえ。
FilesController でもよさそうに見えました。

ポイントとしては、ネーミングをうまくやらないとどこかで読み替えが必要になるということですね。
そして、その読み替えはずっとアプリの中に滞留し続けます。

* `/files` のコントローラーに機能をたそう → FilesController じゃないんだっけ 
* `/files` 以下にさらにコントーラーを配置しよう → `/files/metadata` は Files::MetadataController? Items::MetadataController?
    * すべての routes に as をつけていく?

Rails では適切な名前をつけておくと自然にコード量が少なくなるように作られています。
他の名前をつけることも可能ですが、その場合の特別な書き方をしていく必要があり、レールの隣を徒歩で歩くことになります。

今回はこれ以上広がることはなさそうですし、今から戻ってリネームしようとは言いませんが、ネーミングが重要だというのは覚えておいてください。

a5ob7r replied:
なるほど、必ずしもリソース名とコントローラ名を一致させる必要はないんですね。

tk0miya replied:
言いたいのはモデル名とコントローラー名かな。

だとしたら、そのとおりです。
シンプルなコントローラーでは、ひとつのコントローラーで 1種類のモデルだけを扱いますが、
複雑なコントローラーでは、ひとつのコントローラーで複数のモデルを扱うことは多々あります。

たとえば、SignupController というコントローラーでは、複数のモデルを扱う可能性が高いです。

* アカウントを作る (Account, User, Profile など)
* 決済手段を登録する (CreditCard, Payment, PaymentMethod など)
* 契約をする (Subscription, Contract など)

ですから、モデル名とコントローラー名を一致しないというのは多々あります。
コントローラーは画面や API といった操作を中心としているので、一度の操作で複数のデータを扱う場合は操作名(もしくはそれに準ずるデータ名)をつけることになります。

一方で、シンプルなコントローラーというのもよく使います。
この場合は、可能な限りモデル名とコントローラー名は揃えておいたほうがわかりやすいです。
敢えて別のものを付けることはできますが、検索性は落ちますし、別名をつけるメリットがなければ避けたほうが良いと思います。

a5ob7r replied:
Itemリソース用のスペックを探そうとして、Fileで検索しても見つからず、リソース名がItemであったことを思い出すという見事な伏線回収をしている。

tk0miya commented on spec/support/authorization.rb:
```diff
@@ -0,0 +1,110 @@
+#   let(:request_path) { '/path/to/resource' }
+#   let(:access_token) { an_access_token }
+# end
+RSpec.shared_examples 'OAuth 2.0のResource Owner Password Credentials Grantを利用したBearer Tokenによる認可を行う' do
```
この shared_example は POST 専用であってます?

a5ob7r replied:
先日お話ししましたが、この#shared_examplesは削除しました。

tk0miya commented on spec/requests/items_spec.rb:
```diff
@@ -0,0 +1,63 @@
+      end
+    end
+
+    it_behaves_like 'OAuth 2.0のResource Owner Password Credentials Grantを利用したBearer Tokenによる認可を行う' do
+      let(:request_path) { files_path }
+    end
```
shared_example の内容を見るに、認証に失敗した場合を指していますよね?
context をつけたほうがよいのでは?

tk0miya commented on spec/requests/items_spec.rb:
```diff
@@ -0,0 +1,63 @@
+        it { is_expected.to have_http_status :created }
+        it { is_expected.to have_attributes media_type: 'application/json' }
+
+        it 'アップロードしたファイルのメタデータがボディにJSONとして格納される' do
```
この API の期待値ってこれだけであってます?
データベースに保存されることあたりは仕様外の動きですか?

a5ob7r replied:
書き忘れです。モデルが作成されたことを確認するテストケースを追加しました。

a5ob7r committed:
Add some getter using attr_reader instead of handmade definitions

In Ruby, undefined instance variables return nil if they are referenced.

a5ob7r committed:
Remove unnecessary getters

We don't need to expose them because the users of this concern aren't
interested in them and no need to know about them.

a5ob7r committed:
Remove a unnecessary access_token set method

We don't need to the setter interface (but it's not a real setter...)
because nobody calls it from external of the object.

a5ob7r committed:
Fix the boolean method name

a5ob7r committed:
Add #current_user to hide the not concern from the action

a5ob7r committed:
Move #current_user to the Authorization concern

Controllers are not interested in any access tokens. The concern is only
a user object.

KT-tim commented on app/resources/problem_details/item_resource.rb:
```diff
@@ -0,0 +1,13 @@
+
+    many :errors, resource: ProblemDetails::ActiveModel::ErrorsResource
+
+    def to_hash
```
これはどういう事情で用意してるんですか？

a5ob7r replied:
レスポンスを作成する際に自動的にハッシュに変換させるために実装しています。

https://docs.ruby-lang.org/ja/latest/method/Object/i/to_hash.html

これがない場合

```ruby
render problem: ProblemDetails::ItemResource.new(file)
```

ではなく

```ruby
render problem: ProblemDetails::ItemResource.new(file).to_h
```

とする必要があります。

a5ob7r replied:
ちょうどこの辺りの話題への対処です。

https://github.com/sls-training/2023-API-training-server/pull/30#discussion_r1227879903

KT-tim replied:
なるほどわかりました

以下感想：

内部的にproblemに渡されたオブジェクトのto_hashを呼ぶようになってるんですかね

hashを渡してもいいし、to_hashが生えてる普通のオブジェクトを渡してもいい構造というのはどうなってるんだろう　そもそもhashにto_hashが生えてたので、to_hashが生えてればなんでもいいのかな

a5ob7r replied:
> 内部的にproblemに渡されたオブジェクトのto_hashを呼ぶようになってるんですかね

この辺りで言及されていますね。

> - ProblemDetails::Document ではその他のオプションとして @extensions に保存され、最後に Hash#merge でレスポンスに取り込まれている
https://github.com/nikushi/problem_details/blob/master/lib/problem_details/document.rb#L28
> - Hash#merge のドキュメントには以下の記述がある
othersがハッシュではない場合、othersのメソッドto_hashを使って暗黙の変換を試みます。
https://docs.ruby-lang.org/ja/latest/method/Hash/i/merge.html

https://github.com/sls-training/2023-API-training-server/pull/30#discussion_r1227894536

なので、Ruby、というかそのライブラリの仕様ということになりますかね。

KT-tim replied:
> othersがハッシュではない場合、othersのメソッドto_hashを使って暗黙の変換を試みます。

ほうほう　そういう話だったんですね

tk0miya replied:
`#to_h` は hash-like なオブジェクトに変換するメソッド、 `#to_hash` は Hash オブジェクトに変換するメソッドで微妙に立ち位置が違います。

この記事だと explicit, implicit と紹介されてますね。
https://techracho.bpsinc.jp/hachi8833/2018_06_06/56984

a5ob7r committed:
Remove a shared_examples

The shared_examples has some parameters and it's too complicated to
share test examples and hard to maintenance it.

a5ob7r committed:
Add a missing test case for record creation for file uploading

a5ob7r committed:
Change some before action names for semantics

a5ob7r commented:
修正が必要そうな箇所については対応しました。

tim-toga commented on app/controllers/concerns/authorization.rb:
```diff
@@ -0,0 +1,56 @@
+    before_action :set_token_and_options, :validate_token_and_options
+    before_action :authenticate_access_token
+    before_action :require_authorized_as_readable, only: %i[index show]
+    before_action :require_authorized_as_writable, only: %i[create update destroy]
```
おそらく問題は無いだろうという前提で、
このconcernを使うcontrollerは必ずCRUDなメソッド名をつける必要がありますね。

上以外のメソッド名を使うなというのは分かるのですが、concernsは全てのアクションで共通のものが書かれているという印象があるので、onlyの指定があることになんとなく引っかかっただけです。

tk0miya replied:
基本のアクション名以外を使う場合は、明示的に `require_authorized_as_*` を読んでね、って作りですよね。
割り切りとしてはありだと思いますよ。

デフォルトの設定を持っているので、記述量が減る (自動的に設定してくれる) というメリットがありますよね。
一方で、デフォルト以外のアクションについては明示的に書く必要があるので忘れないか不安、というデメリットもありそうです。

他に思いつく戦略としてはこんな感じかな。

* デフォルトでは権限を付与せず、include 側で明示的に before_action をセットさせる
    * 常にセットさせることで、基本アクションも、その他のアクションもちゃんとセットしてもらう
* create, update, destroy 以外は readable で検証してしまう (`unless: ...`)
    * 必ず何かしら権限チェックが行われる仕組み
    * もし as_writable が指定された場合は readable の検証をスキップする仕組みになってるとよさそう

ここらへんは、利便性と設定漏れのバランスをどこで取るかという concern のデザインがポイントになりますね。
